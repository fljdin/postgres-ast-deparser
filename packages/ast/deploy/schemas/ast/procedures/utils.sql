-- Deploy schemas/ast/procedures/utils to pg

-- requires: schemas/ast/schema

BEGIN;


COMMIT;
