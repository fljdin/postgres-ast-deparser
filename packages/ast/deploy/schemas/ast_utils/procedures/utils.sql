-- Deploy schemas/ast_utils/procedures/utils to pg

-- requires: schemas/ast_utils/schema

BEGIN;

CREATE FUNCTION ast_utils.interval (
  n int
) returns text[] as $$
	select (CASE 
WHEN ( n = 2 ) THEN ARRAY[ 'month' ]
WHEN ( n = 4 ) THEN ARRAY[ 'year' ]
WHEN ( n = 6 ) THEN ARRAY[ 'year', 'month' ]
WHEN ( n = 8 ) THEN ARRAY[ 'day' ]
WHEN ( n = 1024 ) THEN ARRAY[ 'hour' ]
WHEN ( n = 1032 ) THEN ARRAY[ 'day', 'hour' ]
WHEN ( n = 2048 ) THEN ARRAY[ 'minute' ]
WHEN ( n = 3072 ) THEN ARRAY[ 'hour', 'minute' ]
WHEN ( n = 3080 ) THEN ARRAY[ 'day', 'minute' ]
WHEN ( n = 4096 ) THEN ARRAY[ 'second' ]
WHEN ( n = 6144 ) THEN ARRAY[ 'minute', 'second' ]
WHEN ( n = 7168 ) THEN ARRAY[ 'hour', 'second' ]
WHEN ( n = 7176 ) THEN ARRAY[ 'day', 'second' ]
WHEN ( n = 32767 ) THEN ARRAY[]::text[]
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.reserved (
  str text
) returns boolean as $$
	select exists( select 1 from pg_get_keywords() where catcode = 'R' AND word=str  );
$$  
LANGUAGE 'sql' SECURITY DEFINER;

CREATE FUNCTION ast_utils.objtypes ()
returns text[] as $$
	select ARRAY[ 'ACCESS METHOD', 'AGGREGATE', NULL, NULL, NULL, 'CAST', 'COLUMN', 'COLLATION', 'CONVERSION', 'DATABASE', NULL, NULL, 'DOMAIN', 'CONSTRAINT', NULL, 'EXTENSION', 'FOREIGN DATA WRAPPER', 'SERVER', 'FOREIGN TABLE', 'FUNCTION', 'INDEX', 'LANGUAGE', 'LARGE OBJECT', 'MATERIALIZED VIEW', 'OPERATOR CLASS', 'OPERATOR', 'OPERATOR FAMILY', 'POLICY', NULL, NULL, 'ROLE', 'RULE', 'SCHEMA', 'SEQUENCE', NULL, 'STATISTICS', 'CONSTRAINT', 'TABLE', 'TABLESPACE', 'TRANSFORM', 'TRIGGER', 'TEXT SEARCH CONFIGURATION', 'TEXT SEARCH DICTIONARY', 'TEXT SEARCH PARSER', 'TEXT SEARCH TEMPLATE', 'TYPE', NULL, 'VIEW' ]::text[];
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttype_idxs (typ text)
returns int as $$
  select (CASE
WHEN (typ = 'CONSTR_NULL') THEN 0
WHEN (typ = 'CONSTR_NOTNULL') THEN 1
WHEN (typ = 'CONSTR_DEFAULT') THEN 2
WHEN (typ = 'CONSTR_IDENTITY') THEN 3
WHEN (typ = 'CONSTR_CHECK') THEN 4
WHEN (typ = 'CONSTR_PRIMARY') THEN 5
WHEN (typ = 'CONSTR_UNIQUE') THEN 6
WHEN (typ = 'CONSTR_EXCLUSION') THEN 7
WHEN (typ = 'CONSTR_FOREIGN') THEN 8
WHEN (typ = 'CONSTR_ATTR_DEFERRABLE') THEN 9
WHEN (typ = 'CONSTR_ATTR_NOT_DEFERRABLE') THEN 10
WHEN (typ = 'CONSTR_ATTR_DEFERRED') THEN 11
WHEN (typ = 'CONSTR_ATTR_IMMEDIATE') THEN 12
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.constrainttypes (contype int)
returns text as $$
  select (CASE
WHEN (contype =  0 ) THEN 'NULL'
WHEN (contype =  1 ) THEN 'NOT NULL'
WHEN (contype =  2 ) THEN 'DEFAULT'
WHEN (contype =  4 ) THEN 'CHECK'
WHEN (contype =  5 ) THEN 'PRIMARY KEY'
WHEN (contype =  6 ) THEN 'UNIQUE'
WHEN (contype =  7 ) THEN 'EXCLUDE'
WHEN (contype =  8 ) THEN 'REFERENCES'
END);
$$
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.objtypes_idxs (typ text)
returns int as $$
	select (CASE
WHEN (typ = 'OBJECT_ACCESS_METHOD') THEN 0
WHEN (typ = 'OBJECT_AGGREGATE') THEN 1
WHEN (typ = 'OBJECT_AMOP') THEN 2
WHEN (typ = 'OBJECT_AMPROC') THEN 3
WHEN (typ = 'OBJECT_ATTRIBUTE') THEN 4
WHEN (typ = 'OBJECT_CAST') THEN 5
WHEN (typ = 'OBJECT_COLUMN') THEN 6
WHEN (typ = 'OBJECT_COLLATION') THEN 7
WHEN (typ = 'OBJECT_CONVERSION') THEN 8
WHEN (typ = 'OBJECT_DATABASE') THEN 9
WHEN (typ = 'OBJECT_DEFAULT') THEN 10
WHEN (typ = 'OBJECT_DEFACL') THEN 11
WHEN (typ = 'OBJECT_DOMAIN') THEN 12
WHEN (typ = 'OBJECT_DOMCONSTRAINT') THEN 13
WHEN (typ = 'OBJECT_EVENT_TRIGGER') THEN 14
WHEN (typ = 'OBJECT_EXTENSION') THEN 15
WHEN (typ = 'OBJECT_FDW') THEN 16
WHEN (typ = 'OBJECT_FOREIGN_SERVER') THEN 17
WHEN (typ = 'OBJECT_FOREIGN_TABLE') THEN 18
WHEN (typ = 'OBJECT_FUNCTION') THEN 19
WHEN (typ = 'OBJECT_INDEX') THEN 20
WHEN (typ = 'OBJECT_LANGUAGE') THEN 21
WHEN (typ = 'OBJECT_LARGEOBJECT') THEN 22
WHEN (typ = 'OBJECT_MATVIEW') THEN 23
WHEN (typ = 'OBJECT_OPCLASS') THEN 24
WHEN (typ = 'OBJECT_OPERATOR') THEN 25
WHEN (typ = 'OBJECT_OPFAMILY') THEN 26
WHEN (typ = 'OBJECT_POLICY') THEN 27
WHEN (typ = 'OBJECT_PUBLICATION') THEN 28
WHEN (typ = 'OBJECT_PUBLICATION_REL') THEN 29
WHEN (typ = 'OBJECT_ROLE') THEN 30
WHEN (typ = 'OBJECT_RULE') THEN 31
WHEN (typ = 'OBJECT_SCHEMA') THEN 32
WHEN (typ = 'OBJECT_SEQUENCE') THEN 33
WHEN (typ = 'OBJECT_SUBSCRIPTION') THEN 34
WHEN (typ = 'OBJECT_STATISTIC_EXT') THEN 35
WHEN (typ = 'OBJECT_TABCONSTRAINT') THEN 36
WHEN (typ = 'OBJECT_TABLE') THEN 37
WHEN (typ = 'OBJECT_TABLESPACE') THEN 38
WHEN (typ = 'OBJECT_TRANSFORM') THEN 39
WHEN (typ = 'OBJECT_TRIGGER') THEN 40
WHEN (typ = 'OBJECT_TSCONFIGURATION') THEN 41
WHEN (typ = 'OBJECT_TSDICTIONARY') THEN 42
WHEN (typ = 'OBJECT_TSPARSER') THEN 43
WHEN (typ = 'OBJECT_TSTEMPLATE') THEN 44
WHEN (typ = 'OBJECT_TYPE') THEN 45
WHEN (typ = 'OBJECT_USER_MAPPING') THEN 46
WHEN (typ = 'OBJECT_VIEW') THEN 47
END);
$$  
LANGUAGE 'sql' IMMUTABLE;

CREATE FUNCTION ast_utils.getgrantobject (node jsonb)
returns text as $$
DECLARE
  objtype int;
  targtype int;
BEGIN 

  objtype = (node->'objtype')::int;
  IF (node->'targtype') IS NOT NULL THEN
    targtype = (node->'targtype')::int;
  END IF;

  IF (objtype = 0) THEN
    RETURN 'COLUMN';
  ELSIF (objtype = 1) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL TABLES IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'TABLES';
    END IF;
    -- TODO could be a view
    RETURN 'TABLE';
  ELSIF (objtype = 2) THEN
    RETURN 'SEQUENCE';
  ELSIF (objtype = 3) THEN
    RETURN 'DATABASE';
  ELSIF (objtype = 4) THEN
    RETURN 'DOMAIN';
  ELSIF (objtype = 5) THEN
    RETURN 'FOREIGN DATA WRAPPER';
  ELSIF (objtype = 6) THEN
    RETURN 'FOREIGN SERVER';
  ELSIF (objtype = 7) THEN
    IF (targtype = 1) THEN 
      RETURN 'ALL FUNCTIONS IN SCHEMA';
    ELSIF (targtype = 2) THEN 
      RETURN 'FUNCTIONS';
    END IF;
    RETURN 'FUNCTION';
  ELSIF (objtype = 8) THEN
    RETURN 'LANGUAGE';
  ELSIF (objtype = 9) THEN
    RETURN 'LARGE OBJECT';
  ELSIF (objtype = 10) THEN
    RETURN 'SCHEMA';
  ELSIF (objtype = 11) THEN
    RETURN 'TABLESPACE';
  ELSIF (objtype = 12) THEN
    RETURN 'TYPE';
  END IF;

  RAISE EXCEPTION 'BAD_EXPRESSION %', 'GrantObjectType';

END;
$$  
LANGUAGE 'plpgsql' IMMUTABLE;

COMMIT;
