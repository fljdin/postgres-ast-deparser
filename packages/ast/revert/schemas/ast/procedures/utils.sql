-- Revert schemas/ast/procedures/utils from pg

BEGIN;

DROP FUNCTION ast.utils;

COMMIT;
