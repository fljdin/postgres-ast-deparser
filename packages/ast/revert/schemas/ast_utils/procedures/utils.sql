-- Revert schemas/ast_utils/procedures/utils from pg

BEGIN;

DROP FUNCTION ast_utils.utils;

COMMIT;
