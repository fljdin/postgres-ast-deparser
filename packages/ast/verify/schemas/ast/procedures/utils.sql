-- Verify schemas/ast/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast.utils');

ROLLBACK;
