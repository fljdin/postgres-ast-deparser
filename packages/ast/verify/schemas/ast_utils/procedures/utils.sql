-- Verify schemas/ast_utils/procedures/utils  on pg

BEGIN;

SELECT verify_function ('ast_utils.utils');

ROLLBACK;
