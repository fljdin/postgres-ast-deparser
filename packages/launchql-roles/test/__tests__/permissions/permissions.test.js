import { closeConnections, createUser, getConnections } from '../../utils';

let db, conn;

let objs = {};

describe('permissions', () => {
  beforeEach(async () => {
    ({ db, conn } = await getConnections());
    objs.user1 = await createUser(db, undefined, 'user1');
    conn.setContext({
      role: 'authenticated',
      'jwt.claims.user_id': objs.user1.id
    });
    objs.organization1 = await conn.one(
      'SELECT * FROM roles_public.register_organization($1)',
      ['webql']
    );
    objs.permissions = await db.many('SELECT * FROM permissions_public.permission');
    objs.profiles = await db.many('SELECT * FROM permissions_public.profile');
    objs.profiles = objs.profiles.reduce((m, profile) => {
      m[profile.organization_id] = m[profile.organization_id] || {};
      m[profile.organization_id][profile.name] = profile;
      return m;
    }, {});

  });
  afterEach(async () => {
    await closeConnections({ db, conn });
  });
  it('roles', async () => {
    const result = await conn.any(`SELECT pr.name as profile, pm.name as permission FROM permissions_public.profile pr
	JOIN permissions_private.profile_permissions j ON pr.id = j.profile_id
	JOIN permissions_public.permission	pm ON j.permission_id = pm.id;`);
    expect(result).toMatchSnapshot();
  });
});
