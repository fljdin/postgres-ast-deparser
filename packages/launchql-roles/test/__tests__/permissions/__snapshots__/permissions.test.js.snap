// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`permissions roles 1`] = `
Array [
  Object {
    "permission": "Browse Projects",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Projects",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Projects",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Projects",
    "profile": "Owner",
  },
  Object {
    "permission": "Transfer Projects",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Projects",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Secrets",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Secrets",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Secrets",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Secrets",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Secrets",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Content",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Content",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Content",
    "profile": "Owner",
  },
  Object {
    "permission": "Upload Content",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Content",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Content",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Posts",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Posts",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Posts",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Posts",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Posts",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Users",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Users",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Users",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Users",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Users",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Teams",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Teams",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Teams",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Teams",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Teams",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Roles",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Roles",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Roles",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Roles",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Roles",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Role Profiles",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Role Profiles",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Role Profiles",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Role Profiles",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Role Profiles",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Role Settings",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Role Settings",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Role Settings",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Role Settings",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Role Settings",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Invites",
    "profile": "Owner",
  },
  Object {
    "permission": "Read Invites",
    "profile": "Owner",
  },
  Object {
    "permission": "Approve Invites",
    "profile": "Owner",
  },
  Object {
    "permission": "Edit Invites",
    "profile": "Owner",
  },
  Object {
    "permission": "Add Invites",
    "profile": "Owner",
  },
  Object {
    "permission": "Delete Invites",
    "profile": "Owner",
  },
  Object {
    "permission": "Browse Projects",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Projects",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Projects",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Projects",
    "profile": "Administrator",
  },
  Object {
    "permission": "Transfer Projects",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Projects",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Secrets",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Secrets",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Secrets",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Secrets",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Secrets",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Content",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Content",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Content",
    "profile": "Administrator",
  },
  Object {
    "permission": "Upload Content",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Content",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Content",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Posts",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Posts",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Posts",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Posts",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Posts",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Users",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Users",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Users",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Users",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Users",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Teams",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Teams",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Teams",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Teams",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Teams",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Roles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Roles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Roles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Roles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Roles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Role Profiles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Role Profiles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Role Profiles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Role Profiles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Role Profiles",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Role Settings",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Role Settings",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Role Settings",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Role Settings",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Role Settings",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Invites",
    "profile": "Administrator",
  },
  Object {
    "permission": "Read Invites",
    "profile": "Administrator",
  },
  Object {
    "permission": "Approve Invites",
    "profile": "Administrator",
  },
  Object {
    "permission": "Edit Invites",
    "profile": "Administrator",
  },
  Object {
    "permission": "Add Invites",
    "profile": "Administrator",
  },
  Object {
    "permission": "Delete Invites",
    "profile": "Administrator",
  },
  Object {
    "permission": "Browse Projects",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Projects",
    "profile": "Editor",
  },
  Object {
    "permission": "Edit Projects",
    "profile": "Editor",
  },
  Object {
    "permission": "Add Projects",
    "profile": "Editor",
  },
  Object {
    "permission": "Delete Projects",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Secrets",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Secrets",
    "profile": "Editor",
  },
  Object {
    "permission": "Edit Secrets",
    "profile": "Editor",
  },
  Object {
    "permission": "Add Secrets",
    "profile": "Editor",
  },
  Object {
    "permission": "Delete Secrets",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Content",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Content",
    "profile": "Editor",
  },
  Object {
    "permission": "Edit Content",
    "profile": "Editor",
  },
  Object {
    "permission": "Add Content",
    "profile": "Editor",
  },
  Object {
    "permission": "Delete Content",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Posts",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Posts",
    "profile": "Editor",
  },
  Object {
    "permission": "Edit Posts",
    "profile": "Editor",
  },
  Object {
    "permission": "Add Posts",
    "profile": "Editor",
  },
  Object {
    "permission": "Delete Posts",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Invites",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Invites",
    "profile": "Editor",
  },
  Object {
    "permission": "Edit Invites",
    "profile": "Editor",
  },
  Object {
    "permission": "Add Invites",
    "profile": "Editor",
  },
  Object {
    "permission": "Delete Invites",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Users",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Users",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Teams",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Teams",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Roles",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Roles",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Role Profiles",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Role Profiles",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Role Settings",
    "profile": "Editor",
  },
  Object {
    "permission": "Read Role Settings",
    "profile": "Editor",
  },
  Object {
    "permission": "Upload Content",
    "profile": "Editor",
  },
  Object {
    "permission": "Browse Projects",
    "profile": "Author",
  },
  Object {
    "permission": "Read Projects",
    "profile": "Author",
  },
  Object {
    "permission": "Edit Projects",
    "profile": "Author",
  },
  Object {
    "permission": "Add Projects",
    "profile": "Author",
  },
  Object {
    "permission": "Delete Projects",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Secrets",
    "profile": "Author",
  },
  Object {
    "permission": "Read Secrets",
    "profile": "Author",
  },
  Object {
    "permission": "Edit Secrets",
    "profile": "Author",
  },
  Object {
    "permission": "Add Secrets",
    "profile": "Author",
  },
  Object {
    "permission": "Delete Secrets",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Content",
    "profile": "Author",
  },
  Object {
    "permission": "Read Content",
    "profile": "Author",
  },
  Object {
    "permission": "Edit Content",
    "profile": "Author",
  },
  Object {
    "permission": "Add Content",
    "profile": "Author",
  },
  Object {
    "permission": "Delete Content",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Posts",
    "profile": "Author",
  },
  Object {
    "permission": "Read Posts",
    "profile": "Author",
  },
  Object {
    "permission": "Edit Posts",
    "profile": "Author",
  },
  Object {
    "permission": "Add Posts",
    "profile": "Author",
  },
  Object {
    "permission": "Delete Posts",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Invites",
    "profile": "Author",
  },
  Object {
    "permission": "Read Invites",
    "profile": "Author",
  },
  Object {
    "permission": "Edit Invites",
    "profile": "Author",
  },
  Object {
    "permission": "Add Invites",
    "profile": "Author",
  },
  Object {
    "permission": "Delete Invites",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Users",
    "profile": "Author",
  },
  Object {
    "permission": "Read Users",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Teams",
    "profile": "Author",
  },
  Object {
    "permission": "Read Teams",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Roles",
    "profile": "Author",
  },
  Object {
    "permission": "Read Roles",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Role Profiles",
    "profile": "Author",
  },
  Object {
    "permission": "Read Role Profiles",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Role Settings",
    "profile": "Author",
  },
  Object {
    "permission": "Read Role Settings",
    "profile": "Author",
  },
  Object {
    "permission": "Upload Content",
    "profile": "Author",
  },
  Object {
    "permission": "Browse Projects",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Projects",
    "profile": "Contributor",
  },
  Object {
    "permission": "Edit Projects",
    "profile": "Contributor",
  },
  Object {
    "permission": "Add Projects",
    "profile": "Contributor",
  },
  Object {
    "permission": "Delete Projects",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Secrets",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Secrets",
    "profile": "Contributor",
  },
  Object {
    "permission": "Edit Secrets",
    "profile": "Contributor",
  },
  Object {
    "permission": "Add Secrets",
    "profile": "Contributor",
  },
  Object {
    "permission": "Delete Secrets",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Content",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Content",
    "profile": "Contributor",
  },
  Object {
    "permission": "Edit Content",
    "profile": "Contributor",
  },
  Object {
    "permission": "Add Content",
    "profile": "Contributor",
  },
  Object {
    "permission": "Delete Content",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Posts",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Posts",
    "profile": "Contributor",
  },
  Object {
    "permission": "Edit Posts",
    "profile": "Contributor",
  },
  Object {
    "permission": "Add Posts",
    "profile": "Contributor",
  },
  Object {
    "permission": "Delete Posts",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Invites",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Invites",
    "profile": "Contributor",
  },
  Object {
    "permission": "Edit Invites",
    "profile": "Contributor",
  },
  Object {
    "permission": "Add Invites",
    "profile": "Contributor",
  },
  Object {
    "permission": "Delete Invites",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Users",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Users",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Teams",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Teams",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Roles",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Roles",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Role Profiles",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Role Profiles",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Role Settings",
    "profile": "Contributor",
  },
  Object {
    "permission": "Read Role Settings",
    "profile": "Contributor",
  },
  Object {
    "permission": "Upload Content",
    "profile": "Contributor",
  },
  Object {
    "permission": "Browse Projects",
    "profile": "Member",
  },
  Object {
    "permission": "Read Projects",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Secrets",
    "profile": "Member",
  },
  Object {
    "permission": "Read Secrets",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Content",
    "profile": "Member",
  },
  Object {
    "permission": "Read Content",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Posts",
    "profile": "Member",
  },
  Object {
    "permission": "Read Posts",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Users",
    "profile": "Member",
  },
  Object {
    "permission": "Read Users",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Teams",
    "profile": "Member",
  },
  Object {
    "permission": "Read Teams",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Roles",
    "profile": "Member",
  },
  Object {
    "permission": "Read Roles",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Role Profiles",
    "profile": "Member",
  },
  Object {
    "permission": "Read Role Profiles",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Role Settings",
    "profile": "Member",
  },
  Object {
    "permission": "Read Role Settings",
    "profile": "Member",
  },
  Object {
    "permission": "Browse Invites",
    "profile": "Member",
  },
  Object {
    "permission": "Read Invites",
    "profile": "Member",
  },
]
`;
